src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -3827,62 +3821,172 @@ VOID MethodTableBuilder::AllocateFieldDescs()
     }
 }
 
-//*******************************************************************************
-BOOL MethodTableBuilder::IsSelfReferencingStaticValueTypeField(mdToken     dwByValueClassToken,
-                                                               bmtInternalInfo* bmtInternal,
-                                                               const bmtGenericsInfo *bmtGenerics,
-                                                               PCCOR_SIGNATURE pMemberSignature,
-                                                               DWORD       cMemberSignature)
+CorElementType MethodTableBuilder::GetCorElementTypeOfTypeDefOrRefForStaticField(Module* module, mdToken typeDefOrRef)
 {
     STANDARD_VM_CONTRACT;
 
-    if (dwByValueClassToken != this->GetCl())
+    MethodTable *pMTFound = NULL;
+
+    if (TypeFromToken(typeDefOrRef) == mdtTypeDef)
     {
-        return FALSE;
+        pMTFound = module->LookupTypeDef(typeDefOrRef).AsMethodTable();
+    }
+    else if (TypeFromToken(typeDefOrRef) == mdtTypeRef)
+    {
+        pMTFound = module->LookupTypeRef(typeDefOrRef).AsMethodTable();
     }
 
-    if (!bmtGenerics->HasInstantiation())
+    if ((pMTFound == NULL) && bmtInternal->pType != NULL && (typeDefOrRef != GetCl()))
```

Comment:
The GetCl check here seems like it works for exact self-references, but it might not be enough for self-references via interfaces. (I don't think that's necessarily a problem)

Reply:
Indeed. This is intended to reduce the number of exceptions throw, not affect the correctness. I'll add a comment to that effect.